ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallViewModel.kt

```diff
@@ -36,15 +45,49 @@ internal class PaywallViewModelImpl(offering: Offering?) : ViewModel(), PaywallV
         }
     }
 
-    override fun purchasePackage(activity: Activity, packageToPurchase: Package) {
+    override fun selectPackage(packageToSelect: Package) {
+        _state.value = when (val currentState = _state.value) {
+            is PaywallViewState.Template2 -> {
+                currentState.copy(selectedPackage = packageToSelect)
+            }
+            else -> {
+                Log.e(TAG, "Unexpected state trying to select package: $currentState")
+                currentState
+            }
+        }
+    }
+
+    override fun purchaseSelectedPackage(activity: Activity) {
+        when (val currentState = _state.value) {
+            is PaywallViewState.Template2 -> {
+                purchasePackage(activity, currentState.selectedPackage)
+            }
+            else -> {
+                Log.e(TAG, "Unexpected state trying to purchase package: $currentState")
+            }
+        }
+    }
+
+    override fun restorePurchases() {
+        viewModelScope.launch {
+            try {
+                val customerInfo = Purchases.sharedInstance.awaitRestore()
+                Log.i(TAG, "Restore purchases successful: $customerInfo")
+            } catch (e: PurchasesException) {
+                Log.e(TAG, "Error restoring purchases: $e")
+            }
+        }
+    }
+
+    private fun purchasePackage(activity: Activity, packageToPurchase: Package) {
         viewModelScope.launch {
             try {
                 val purchaseResult = Purchases.sharedInstance.awaitPurchase(
                     PurchaseParams.Builder(activity, packageToPurchase).build(),
                 )
-                Log.i("PaywallTester", "Purchased package: ${purchaseResult.storeTransaction}")
+                Log.i(TAG, "Purchased package: ${purchaseResult.storeTransaction}")
             } catch (e: PurchasesException) {
-                Log.e("PaywallTester", "Error purchasing package: $e")
+                Log.e(TAG, "Error purchasing package: $e")
```

Comment:
We should probably improve our logging mechanisms for the new library soon, to avoid this repetition and provide more details.

Reply:
Done in #1233